### PR TITLE
Locks and Multi fan-in and fan-out

### PR DESCRIPTION
**Pull request by Suhas-G**
Issues:
Locks: When multiple writers are writing to same wire, its necessary to lock the shared memory while write is in progress.
Enable in FSM: In the case of FSMs, a block should be able to disable itself after it transitions to the next state. But the Enable wire of a block is an input block and not writeable by it. This PR changes this by giving a separate interface for enable wire as a property of the inputs block. And also allows the block to disable by setting `inputs.enable = False` to disable itself safely. 
Multi Fan-in and Fan-out: When we have the scenario of multiple wires going to same input or same output having multiple wires to different inputs, we have to use the same underlying shared memory so that data can be modified correctly.

### CODE REVIEW

**Review by toshan-luktuke — CHANGES_REQUESTED** [review]
Tested with:
- [x] Multiple inputs (2 - 3 wires coming in a single block)
- [x] Multiple outputs (2 - 3 wires going out of a single block)

Remaining:
- [ ] FSM Testing with Enable

File: backend/staticfiles/synthesis/main.py [hunk 1]
```diff
@@ -30,6 +31,13 @@ def clean_shared_memory(names):
         except FileNotFoundError:
             pass
 
+    # Release all locks
+    for name in names:
+        try:
+            names[name]["lock"].release()
```
[the remark this answers is not in the document]
Comment (Suhas-G) [reply backend/staticfiles/synthesis/main.py:37]: Yes. Thanks. I'll change that.

**Comment by toshan-luktuke** [comment]
I'm not able to figure out how the `enable` implementation is intended to work, I got it working but I had to make a few changes to the code here and there.
If possible can you give an example of the intended way to use it?

**Comment by Suhas-G** [comment]
Did the fsm structure with enable I showed work? Shall I merge this PR?

**Comment by toshan-luktuke** [comment]
Let's merge `outputs.py` and `main.py`. I've made some modifications to `inputs.py` to write to the Shared Memory like we discussed. Its better if I make a separate PR for that

**Review by toshan-luktuke — APPROVED** [review]

**Pull request merged**